### PR TITLE
[Serverless] Support sampling in universal instrumentation

### PR DESCRIPTION
**Pull request by maxday**
### What does this PR do?

Support sampling in universal instrumentation

### Motivation

Feature parity

### Describe how to test/QA your changes

Deployed a function with DD_TRACE_SAMPLE_RATE = 0.1
Out of 1065 invocations, only 154 traces were generated (0.14%)
Screen Shot 2022-04-28 at 9 52 26 AM
Screen Shot 2022-04-28 at 9 51 57 AM

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/serverless/invocationlifecycle/lifecycle.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ type LifecycleProcessor struct {
 	ProcessTrace        func(p *api.Payload)
 	Demux               aggregator.Demultiplexer
 	DetectLambdaLibrary func() bool
+	SamplingRate        float64
```
Comment (jcstorms1) [inline pkg/serverless/invocationlifecycle/lifecycle.go:23]: Nit: Should this be an `int32`? That is what trace chunks use and the value can only be 0, 1, or 2 as far as I'm aware.

**Review by jcstorms1 — APPROVED** [review]
💯

**Comment by hghotra** [comment]
Spoke with @maxday, we should modify the logic to:
- extract the trace context from the event headers, and pass those to the tracer to use for the execution span
- let the tracer create the `trace id`, `parent id` & `sampling-priority` when no trace context provided by upstream service

This will make sure the sampling decision is made by the tracer based on `DD_TRACE_SAMPLE_RATE` or `DD_TRACE_SAMPLING_RULES` and we don't duplicate that effort in the extension.

File: pkg/serverless/invocationlifecycle/trace.go [hunk 2]
```diff
@@ -139,6 +148,21 @@ func convertStrToUnit64(s string) (uint64, error) {
...
+		samplingPriority = sampler.SamplingPriority(v)
+	} else {
+		// try to look for direction invocation headers
+		if v, err := strconv.ParseInt(directInvokeHeader, 10, 8); err == nil {
+			samplingPriority = sampler.SamplingPriority(v)
+		}
+	}
+	return samplingPriority
+}
+
```
[the remark this answers is not in the document]
Comment (maxday) [reply pkg/serverless/invocationlifecycle/trace.go:165]: This method is still useful to parse the current existing sampling decision. It won't create a new one, but we still need to extract it. Does it make sense?

[the remark this answers is not in the document]
Comment (hghotra) [reply pkg/serverless/invocationlifecycle/trace.go:165]: Yeah, that makes sense

**Review by hghotra — APPROVED** [review]
👍🏼

File: pkg/serverless/invocationlifecycle/lifecycle.go [hunk 3]
```diff
@@ -56,9 +56,7 @@ func (lp *LifecycleProcessor) OnInvokeEnd(endDetails *InvocationEndDetails) {
 
 		if lp.InferredSpansEnabled {
 			log.Debug("[lifecycle] Attempting to complete the inferred span")
-			inferredspan.CompleteInferredSpan(
-				lp.ProcessTrace, endDetails.EndTime, endDetails.IsError, inferredSpan,
-			)
+			inferredspan.CompleteInferredSpan(lp.ProcessTrace, endDetails.EndTime, endDetails.IsError, inferredSpan, TraceID(), SamplingPriority())
```
Comment (jcstorms1) [inline pkg/serverless/invocationlifecycle/lifecycle.go:59]: 💯

**Review by jcstorms1 — APPROVED** [review]

**Pull request merged**